### ISSUE TITLE
AoT error with interface in type of @Component's constructor parameters

### ISSUE THREAD

**Issue by victornoel** [angular/angular#15424]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report
```

**Current behavior**

Compiling with `--prod` of the example below gives me something like this:
```sh
$ ng serve --prod
****************************************************************************************
This is a simple server for use in testing or debugging Angular applications locally.
It hasn't been reviewed for security issues.

DON'T USE IT FOR PRODUCTION USE!
****************************************************************************************
** NG Live Development Server is running on http://localhost:4200 **
 10% building modules 3/3 modules 0 activeFailed on type {"filePath":"/home/vnoel/Linagora/Petals/dev/git/bug/src/app/app.component.ts","name":"BugComponent","members":[]} with error Error: Error encountered resolving symbol values statically. Expression form not supported (position 20:18 in the original .ts file), resolving symbol BugComponent in /home/vnoel/Linagora/Petals/dev/git/bug/src/app/app.component.ts
Hash: 0963bfd57394d0f63877                                                               
Time: 8008ms
chunk    {0} polyfills.3378afd5b9dc00772216.bundle.js (polyfills) 157 kB {4} [initial] [rendered]
chunk    {1} main.c888f1991d9ebf9e24f9.bundle.js (main) 680 bytes {3} [initial] [rendered]
chunk    {2} styles.d41d8cd98f00b204e980.bundle.css (styles) 69 bytes {4} [initial] [rendered]
chunk    {3} vendor.6b1b7c43943b5a5fe6d7.bundle.js (vendor) 1.12 MB [initial] [rendered]
chunk    {4} inline.f5af5e469e05aa49837b.bundle.js (inline) 0 bytes [entry] [rendered]

ERROR in Error encountered resolving symbol values statically. Expression form not supported (position 20:18 in the original .ts file), resolving symbol BugComponent in /home/vnoel/Linagora/Petals/dev/git/bug/src/app/app.component.ts

ERROR in ./src/main.ts
Module not found: Error: Can't resolve './$$_gendir/app/app.module.ngfactory' in '/home/vnoel/Linagora/Petals/dev/git/bug/src'
 @ ./src/main.ts 4:0-74
 @ multi webpack-dev-server/client?http://localhost:4200 ./src/main.ts
webpack: Failed to compile.

```

**Expected behavior**

No error :)

**Minimal reproduction of the problem with instructions**
To reproduce:
```sh
$ mkdir bug
$ cd bug
$ ng new
$ cat << EOF | git apply
diff --git a/src/app/app.component.ts b/src/app/app.component.ts
index ff63e05..ddbf7a1 100644
--- a/src/app/app.component.ts
+++ b/src/app/app.component.ts
@@ -1,4 +1,4 @@
-import { Component } from '@angular/core';
+import { Component, Inject } from '@angular/core';

 @Component({
   selector: 'app-root',
@@ -8,3 +8,15 @@ import { Component } from '@angular/core';
 export class AppComponent {
   title = 'app works!';
 }
+
+export interface IInterface { }
+
+@Component({
+  selector: 'app-component',
+  template: '<div></div>'
+})
+export class BugComponent {
+  constructor(
+    public data: { prop: IInterface }
+    ) { }
+}
diff --git a/src/app/app.module.ts b/src/app/app.module.ts
index 67ae491..d43010b 100644
--- a/src/app/app.module.ts
+++ b/src/app/app.module.ts
@@ -3,11 +3,12 @@ import { NgModule } from '@angular/core';
 import { FormsModule } from '@angular/forms';
 import { HttpModule } from '@angular/http';

-import { AppComponent } from './app.component';
+import { AppComponent, BugComponent } from './app.component';

 @NgModule({
   declarations: [
-    AppComponent
+    AppComponent,
+    BugComponent
   ],
   imports: [
     BrowserModule,
EOF
$ ng serve --prod
```

* **Angular version:** 4.0.0-rc.6 / cli 1.0.0-rc.4

* **Language:** TypeScript

* **Node (for AoT issues):** `node --version` = v6.10.0

**Comment by DzmitryShylovich** [comment]
Please add a small repo that we can clone

**Comment by victornoel** [comment]
@DzmitryShylovich you were right to ask, the problem is not the one I thought it was.

Simply add such a component:
```
export interface IInterface { }

@Component({
  selector: 'app-component',
  template: '<div></div>'
})
export class BugComponent {
  constructor(
    public data: { prop: IInterface }
    ) { }
}
```

to an existing project and add the component to the `declarations` of a `NgModule` and it will show the bug.

To reproduce:
```sh
$ mkdir bug
$ ng new bug
$ cat << EOF | git apply
diff --git a/src/app/app.component.ts b/src/app/app.component.ts
index ff63e05..ddbf7a1 100644
--- a/src/app/app.component.ts
+++ b/src/app/app.component.ts
@@ -1,4 +1,4 @@
-import { Component } from '@angular/core';
+import { Component, Inject } from '@angular/core';

 @Component({
   selector: 'app-root',
@@ -8,3 +8,15 @@ import { Component } from '@angular/core';
 export class AppComponent {
   title = 'app works!';
 }
+
+export interface IInterface { }
+
+@Component({
+  selector: 'app-component',
+  template: '<div></div>'
+})
+export class BugComponent {
+  constructor(
+    public data: { prop: IInterface }
+    ) { }
+}
diff --git a/src/app/app.module.ts b/src/app/app.module.ts
index 67ae491..d43010b 100644
--- a/src/app/app.module.ts
+++ b/src/app/app.module.ts
@@ -3,11 +3,12 @@ import { NgModule } from '@angular/core';
 import { FormsModule } from '@angular/forms';
 import { HttpModule } from '@angular/http';

-import { AppComponent } from './app.component';
+import { AppComponent, BugComponent } from './app.component';

 @NgModule({
   declarations: [
-    AppComponent
+    AppComponent,
+    BugComponent
   ],
   imports: [
     BrowserModule,
EOF
```

**Comment by DzmitryShylovich** [comment]
Yeah, I can reproduce it. But the error message indicated that `{ prop: IInterface }` this form is not supported. btw how do you provide data?

**Comment by victornoel** [comment]
@DzmitryShylovich in my real use case, the component's constructor parameter is declared like this:
```ts
constructor(
    public dialogRef: MdDialogRef<WorkspaceDeleteDialogComponent>,
    // TODO add some type for data when https://github.com/angular/angular/issues/15424 is fixed
    @Inject(MD_DIALOG_DATA) public data: any
  ) { }
```

And I create the dialog like this:
```ts
    this.dialog.open(WorkspaceDeleteDialogComponent, {
      data: { workspace$: this.workspace$ }
    })
```

And `@angular/material` takes care of injecting `data` into my parameter.

**Comment by DzmitryShylovich** [comment]
>  @Inject(MD_DIALOG_DATA) public data: any

does it work without this constructor parameter?

**Comment by victornoel** [comment]
@DzmitryShylovich sorry, I made a mistake in my last comment, I meant the problem was with:
```ts
constructor(
    public dialogRef: MdDialogRef<WorkspaceDeleteDialogComponent>,
    @Inject(MD_DIALOG_DATA) public data: { workspace$: Observable<IWorkspace>} 
  ) { }
```

It does works if I use `@Inject(MD_DIALOG_DATA) public data: any` since it does not have an interface to type the parameter.

**Comment by DzmitryShylovich** [comment]
create an interface
```
export interface DialogData {
  workspace$: Observable<IWorkspace>
}

@Inject(MD_DIALOG_DATA) public data: DialogData 
```

**Comment by victornoel** [comment]
@DzmitryShylovich It does work, thanks for the workaround!

**Comment by DzmitryShylovich** [comment]
>  @Inject(MD_DIALOG_DATA) public data: { workspace$: Observable<IWorkspace>} 

@chuckjaz are there any plans to support this?

**Comment by chuckjaz** [comment]
@DzmitryShylovich This is a bug. We should consider fixing this in a patch release of 4.0.

**Comment by morninng** [comment]
I have an same issue with firebase injection

```
import {AngularFire, FirebaseApp} from 'angularfire2';
@Injectable()
export class LoggerService {
  constructor(@Inject(FirebaseApp) firebaseApp: firebase.app.App) {
   const connectedRef = firebaseApp.database().ref(".info/connected");
```

This code is used for firebase function which is not suported by Angularfire2.
Is there also work around way for it?

**Comment by chuckjaz** [comment]
@morninng, this issue tracker is not suitable for support requests, please repost your issue on [StackOverflow](http://stackoverflow.com/) using tag `angular`.

If you are wondering why we don't resolve support issues via the issue tracker, please [check out this explanation](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-got-a-question-or-problem).

**Comment by morninng** [comment]
@chuckjaz
 Thank you for your pointing out.
I maybe able to wait for your fix as I can build by " ng b --prod --aot=false" as of now
But I will post it on stack over flow, too.

**Comment by victornoel** [comment]
@chuckjaz I don't think this bug is solved, if I write something like this (see the original post for context, what matters is that there is some type for data):
```
@Inject(MD_DIALOG_DATA) public data: { bus: IBusWithContainers }
```

I still get the aforementioned error:
```
ERROR in Error encountered resolving symbol values statically. Expression form not supported (position 135:42 in the original .ts file), resolving symbol BusDeleteDialogComponent in /builds/linagora/petals-cockpit/frontend/src/app/features/cockpit/workspaces/petals-content/petals-bus-view/petals-bus-view.component.ts
```

I'm using now:
```
@angular/cli: 1.1.2
node: 8.1.2
os: linux x64
@angular/animations: 4.1.3
@angular/common: 4.1.3
@angular/compiler: 4.1.3
@angular/core: 4.1.3
@angular/flex-layout: 2.0.0-beta.8
@angular/forms: 4.1.3
@angular/http: 4.1.3
@angular/material: 2.0.0-beta.7
@angular/platform-browser: 4.1.3
@angular/platform-browser-dynamic: 4.1.3
@angular/router: 4.1.3
@angular/cli: 1.1.2
@angular/compiler-cli: 4.1.3
@angular/language-service: 4.1.3
```

**Comment by chuckjaz** [comment]
This bug fix was not cherry-picked into 4.1. It is only available in 4.2 and greater.

**Comment by victornoel** [comment]
@chuckjaz ok thanks, sorry for the confusion :)
